Update style guidelines for API object capitalization & formatting (#23497)
kubernetes/website: content/en/docs/contribute/style/style-guide.md

## Documentation formatting standards

> ### When to use camel case for API objects

Reviewer: Simpler and more direct:
```suggestion
### Use upper camel case for API objects
```

> When you are specifically referring to interacting with an API object, use [camel case](https://en.wikipedia.org/wiki/Camel_case). When you are generally discussing an API object, use sentence capitalization.

Reviewer: Given that many K8s contributors speak English as a second language, we can't take for granted that the meaning of sentence capitalization is obvious. Can you recommend a resource we can link to for more information?

```suggestion
When you refer specifically to interacting with an API object, use [camel case](https://en.wikipedia.org/wiki/Camel_case). When you are generally discussing an API object, use [sentence capitalization]().
```

Author: Will the community oust me for referring to a Microsoft document? ;) 

https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization

Reviewer 2: Why not shorten this to: "When you refer to an API object," ?

Also, the Kubernetes community's documentation has embraced an ambiguous definition of "camel case." The objects in question are capitalized in the first letter and in any subsequent first letters of the closed compound (PodList, e.g.). This is sometimes referred to as "camel case", but when referred to as "Pascal case" is unambiguous. I believe we should standardize on this and describe API objects as being set in Pascal case.

It is at least as accurate, and a lot less confusing.

> When you refer specifically to interacting with an API object, use [CamelCase](https://en.wikipedia.org/wiki/Camel_case). When you are generally discussing an API object, use [sentence-style capitalization](https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization).

Reviewer: ```suggestion
When you refer specifically to interacting with an API object, use [UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case), also known as Pascal Case. When you are generally discussing an API object, use [sentence-style capitalization](https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization).
```

> When you refer specifically to interacting with an API object, use [UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case), also known as Pascal Case. When you are generally discussing an API object, use [sentence-style capitalization](https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization).

Reviewer 3: > When you are generally discussing an API object, use [sentence-style capitalization](https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization).

This isn't quite the approach I usually take. I use sentence-style capitalization **only** when the `kind` of the resource is also an word or bit of IT jargon in the target localization.

**Yes:**
- secret (the English word “secret”)
- namespace (the English word “namespace”)
- Namespace (meaning the `kind` of an object in the API)
- nœud (_in French text_)
- node (_in English text_)
- Node (in any localization - and, again, meaning the `kind` of an object in the API)
- PersistentVolumeClaim (again meaning the `kind` of an object in the API)
- Secret (again meaning the `kind` of an object in the API)
- volume (because you can't `kubectl get Volume` but you can, eg, `kubectl get Secret`)
- certificate signing request (common IT jargon outside of a Kubernetes context)
- CertificateSigningRequest (again, meaning the `kind` of an object in the API)
- CSR (might need expanding if the first use, but subsequent uses are fine)
- PVC (might need expanding if the first use, but subsequent uses are fine)

**No:**
- configmap
- poddisruptionbudget
- tokenreview
- persistentvolume
- persistentVolume (should be PersistentVolume)
- statefulset

I also **advise against** _phrases_ (multiple words) that exactly match the way you would say an API name:
- pod disruption budget
- persistent volume claim
- config map (even “configuration map” would be a bit unidiomatic)
- subject access review
- stateful set
- daemon set

However, because it's a different pronunciation, I think it's fine to write:
- &hellip;implements horizontal Pod autoscaling&hellip;
- Kubernetes persistent storage&hellip;
- &hellip;according to the hisruption budget&hellip;
- &hellip;review subject access to&hellip;
- &hellip;signing the CSR&hellip;

It's not 100% clear cut. There are nuances. If you want to write, eg, “a manual token review process would take too much time” then I would never suggest rewriting that because “TokenReview” sounds the same. _The whole point of these different capitalization styles is to help the reader tell these cases apart_.

{{< table caption = "Do and Don't - API objects" >}}
Do | Don't
:--| :-----

> The pod has two containers. | The Pod has two containers.

Reviewer: 👍

Reviewer 4: 👍

> The deployment is responsible for ... | The Deployment object is responsible for ...

Reviewer 5: See? Here is where the confusion comes.
When you say "the deployment is responsible for ...", I have no clear idea what you are talking about. However, when you say "the Deployment is removed", I can clearly understand wha you mean.

Reviewer: Agreed that "deployment" here may be both consistent to a convention and unhelpful to readers.

@jimangel @liggitt 👋 Any input here?

Reviewer 6: I agree the "Do" example proposed here is confusing. I typically refer to API objects using the kind of the object (`Deployment`, `HorizontalPodAutoscaler`, etc), and I usually monospace it to make it clear it is a programmatic name. The previous guidance seems easier to apply consistently and less likely to lead to confusing constructions like this one.

Was there a problem with the previous guidance this is trying to correct? Maybe an additional example line in this table for `HorizontalPodAutoscaler` would be good to make sure the guidance we come up with works for the range of object names we'll need to apply it to.

Reviewer 3: HorizontalPodAutoscaler is a good example. Some others, that in English are also sentence fragments and a good choice for the natural-language description of what the resource represents:
- CustomResourceDefinition
- PodSecurityPolicy
- PodDisruptionBudget

Reviewer 7: In the DO example above, I would read that as referring to a general use of the term deployment. (There is a rule later talking about using monospaced font to denote a specific `Deployment` object, that clears things up, imo.) I would then think that the DON'T example, would have two problems, Capitalized, like an object, AND using the word "object".  In our docs I've seen more uses of the generic term deployment.

Author: Maybe it's just an example problem like you guys have mentioned. We should clarify those. In general, when I read this page and apply the new formatting, it makes sense and I don't see any confusing references in real world scenario:

https://kubernetes.io/docs/concepts/workloads/controllers/deployment/

Author: Ok, I've updated the examples to ensure every scenario is covered and used something different than deployment as an example. If you also think we need a deployment example that makes more sense than the original, I can add another row to the table.

Reviewer 2: This works right up until someone copies your documentation and pastes it somewhere else. The only reliable means of describing an API object that can be understood unambiguously to be an object and can be read by a non-technical/novice reader (including a technical editor!) to be an object is to identify the object as an object in the text. 

This seems like overkill in the case of Pascal-cased objects such as those Tim mentions, but gets pretty hairy when there's no way to divine whether an object name is Pascal-cased or simply incorrectly capitalized. In other words, you need a standard that works even when someone who doesn't understand your capitalization rules gets their hands on the documentation.

Reviewer: @wabernatScality

> This works right up until someone copies your documentation and pastes it somewhere else. 

This is a use case over which we have almost zero control. Optimizing for it seems like a lot of complication for a small impact.

> A PodList is a list of pods. | A Pod List is a list of pods.

Reviewer: 👍

> A PodList is a list of pods. | A Pod List is a list of pods.

Reviewer 3: In the API, a PodList _is_ a list of Pods.

Author: I wasn't saying that's not the case. I'm just saying that the example was confusing, so trying to use one that is more clear to  show our meaning for when to capitalize vs not.

Reviewer 8: Hello @cmastr .
I see that there is at least one remaining question to address _before_ these changes can merge.
Not to belabor the point but perhaps line 60 is not a fair example and should be removed?

`A PodList is a list of pods. | A Pod List is a list of pods.`

For example (of course the field names would use code style not double quotes):
`A ServiceAccount has twenty-nine fields such as "kind", "version", ...` versus `A Service account is used throughout your Kubernetes documentation` OR
`This page teaches you how to set the "kind" field in a Pod spec` versus `This page teaches you about a Pod spec`

I have not read your Kubernetes documentation, but am guessing that you are looking for some standardization of the Kubernetes terms so that you can more easily refer to this content.

Author: I agree that this example is not good. I was trying not to change so much other than the capitalization which was the intent of this PR. But the examples do definitely need some work. I'm good with removing line 60. Replacing it with something else that is more explanatory of what we're trying to get across. Will update.

Author: Can we do a follow on PR for this item? I just finally figured out how to squash my commits and don't want to add to them now.

Reviewer 8: Hi @cmastr. I am not sure what happened when you "squashed" your commits.
It looks like you have 6 commits for this pull request?
At the beginning of the PR you can see the commits listed and the files changed.
To move forward, you could try and squash again or open a new PR with your updated changes?

Use angle brackets for placeholders. Tell the reader what a placeholder
represents.

> 1. Display information about a pod:

Reviewer: 👍

Reviewer 4: :+1:

## Inline code formatting

> ### Use code style for inline code, commands, and API objects

Reviewer: Adding code styling to API objects would match GitHub's and Google's API conventions, which seems like a trivial way to ease friction. 👍

Reviewer 3: @cmastr, would you be willing to open an issue that recommends a change to use code styling for API objects?
If you do, you can link this PR to the issue.

The reason I'm suggesting that switch is because, firstly, I want to chip in on the discussion with some comments about implemention details, and secondly, I want to separate the “should we do this?” discussion (the issue) from the “is this a good way to do it?” discussion (which can be this pull request and potentially other pull requests if a different person suggests another approach).

All of this is in aid of a low-friction route to revising and improving what seems like a contentious part of the style guide.

Author: Yes sure. Let me see if I can figure out how to open an issue and describe it properly.

Reviewer 8: > The reason I'm suggesting that switch is because, firstly, I want to chip in on the discussion with some comments about implemention details, and secondly, I want to separate the “should we do this?” discussion (the issue) from the “is this a good way to do it?” discussion (which can be this pull request and potentially other pull requests if a different person suggests another approach).
> 
> All of this is in aid of a low-friction route to revising and improving what seems like a contentious part of the style guide.

Yes, create an issue to outline any proposed changes.

Author: Here you go: #23527

Reviewer 8: Hi. I probably would hold off on changing the heading.
I'd like more information about styling API objects before agreeing to this change (provide links to other
style guides, especially guides that reference Kubernetes concepts).

Reviewer: @kbhawkey Google and Github style API objects in monospace font. I think the heading is fine and shouldn't be a blocker.

For inline code in an HTML document, use the `<code>` tag. In a Markdown
document, use the backtick (`` ` ``).

> {{< table caption = "Do and Don't - Use code style for inline code and commands" >}}

Reviewer 3: Does this caption need to change?

Do | Don't
:--| :-----

> The `kubectl run`command creates a pod. | The "kubectl run" command creates a pod.

Reviewer: 👍

Reviewer 3: Not so sure on this one. The `kubectl run`command creates a Pod in the API server. It might also set up a set of containers on a node (a pod), but it might not - for example, if all your nodes are unhealthy you get the Pod object but not the running pod that you asked for.

Reviewer 8: I agree, this particular sentence can be unclear.

Reviewer: The sentence (which is legacy content) may be unclear, but the guidelines about styling are clearer, which fits the scope of this PR and the issue it resolves (#23527). 

I recommend proceeding with this change and clarifying the sentence in a separate PR.

> The `kubectl run`command creates a pod. | The "kubectl run" command creates a pod.

Reviewer 3: This looks wrong: the “Do” example **and** the caption don't match the revised heading.

Perhaps @cmastr you meant to propose changes along the lines of:
```suggestion
The `kubectl run` command creates a `Pod`. | The "kubectl run" command creates a pod.
The kubelet on each node acquires a `Lease`… | The kubelet on each node acquires a lease…
A `PersistentVolume` represents durable storage… | A Persistent Volume represents durable storage…
```

### Separate commands from output

> Verify that the pod is running on your chosen node:

Reviewer: 👍

:--| :-----
To create a ReplicaSet, ... | In order to create a ReplicaSet, ...
See the configuration file. | Please see the configuration file.

> View the pods. | With this next command, we'll view the pods.

Reviewer: 👍

:--| :-----
Version 1.4 includes ... | In version 1.4, we have added ...
Kubernetes provides a new feature for ... | We provide a new feature ...

> This page teaches you how to use pods. | In this page, we are going to learn about pods.

Reviewer: 👍
